### PR TITLE
ENH: Implement arrow support for read_csv with engine=c

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Not really sure if this is the best way forward, but want to see if ci passes everywhere

### CODE REVIEW

File: pandas/_libs/parsers.pyx [hunk 1]
```diff
@@ -341,6 +342,7 @@ cdef class TextReader:
         bint use_nullable_dtypes
         object usecols
         set unnamed_cols  # set[str]
+        object dtype_backend  # str
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx:345]: can this be typed as str?

File: pandas/_libs/parsers.pyx [hunk 2]
```diff
@@ -1440,6 +1448,13 @@ def _maybe_upcast(arr, use_nullable_dtypes: bool = False):
         if use_nullable_dtypes:
             arr = StringDtype().construct_array_type()._from_sequence(arr)
 
+    if use_nullable_dtypes and dtype_backend == "pyarrow":
+        import pyarrow as pa
+        if isinstance(arr, IntegerArray) and arr._mask.all():
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx:1453]: can this be arr.isna().all()?

File: pandas/io/parsers/c_parser_wrapper.py [hunk 3]
```diff
@@ -79,6 +82,10 @@ def __init__(self, src: ReadCsvBuffer[str], **kwds) -> None:
             kwds.pop(key, None)
 
         kwds["dtype"] = ensure_dtype_objs(kwds.get("dtype", None))
+        dtype_backend = get_option("mode.dtype_backend")
+        kwds["dtype_backend"] = dtype_backend
+        if dtype_backend == "pyarrow":
+            import_optional_dependency("pyarrow")  # Ensure that we fail here
```
Comment (jbrockmendel) [inline pandas/io/parsers/c_parser_wrapper.py:88]: "Ensure that we fail here" means "in the case where pyarrow is not installed, better to fail loudly here than somewhere else"?

Comment (phofl) [reply pandas/io/parsers/c_parser_wrapper.py:88]: Yep, will clarify the comment

**Comment by jbrockmendel** [comment]
is there any scenario in which the user might want our cython engine rather than the pyarrow engine?

**Comment by phofl** [comment]
Yes, the arrow engine basically does not support any keywords. c is way more feature complete

File: pandas/_libs/parsers.pyx (hunk 2 above)
Comment (phofl) [reply pandas/_libs/parsers.pyx:1453]: Yep works as well

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**